lib/miou_queue.ml

```diff
@@ -117,18 +116,13 @@ let iter ~f t =
   in
   go head
 
-let rec drop ~f t =
-  let head, tail = snapshot t in
-  if Atomic.compare_and_set t.head head tail then (
-    let rec go prev =
-      if prev != tail then
-        match Atomic.get prev.next with
-        | None -> ()
-        | Some next -> f next.value; go next
-    in
-    go head;
-    tail.value <- Obj.magic ())
-  else drop ~f t
+let rec drop t =
+  let ((head, tail) as snapshot) = snapshot t in
+  if Atomic.compare_and_set t.head head tail
+  then snapshot else drop t
+
+let drop ~f t = iter ~f (drop t)
+let iter ~f t = iter ~f (snapshot t)
```

Comment:
Note that `iter` is not concurrency safe.  If one thread calls `iter ~f q` and another thread performs `dequeue q`, then the `~f` function may be called with `Obj.magic ()`: https://github.com/robur-coop/miou/blob/9d18413ddb739d512210c4a91b8199a403465528/lib/miou_queue.ml#L63

Reply:
More precisely, this is what I encountered on my project. I suspected that this version will help me but it's not - and it's why I decided to switch to a mutex + a simple queue. If you have an idea to have an atomic `iter`, it will be superb but my brain is actually focus and limited on some others stuffs.